### plot_panel_summary.R ###########################################################################
# Plot data summary across targeted regions

### FUNCTIONS ######################################################################################
# function to generate a standardized filename
generate.filename <- function(project.stem, file.core, extension, include.date = TRUE) {

	# build up the filename
	file.name <- paste(project.stem, file.core, sep = '_');
	file.name <- paste(file.name, extension, sep = '.');

	if (include.date) {
		file.name <- paste(Sys.Date(), file.name, sep = '_');
		}

	return(file.name);
	}

# function to write session profile to file
save.session.profile <- function(file.name) {

	# open the file
	sink(file = file.name, split = FALSE);

	# write memory usage to file
	cat('### MEMORY USAGE ###############################################################');
	print(proc.time());

	# write sessionInfo to file
	cat("\n### SESSION INFO ###############################################################");
	print(sessionInfo());

	# close the file
	sink();

	}

# function to trim sample IDs
simplify.ids <- function(x) {
	match <- TRUE;
	if (length(x) == 1) {
		match <- FALSE;
		new.ids <- x;
		}
	index <- 1;
	while (match) {
		if (length(unique(sapply(x,function(i) { unlist(strsplit(i,''))[index] } ))) == 1) {
			index <- index + 1;
			} else {
			new.ids <- sapply(x,function(i) { substring(i,index,nchar(i)) } );
			match <- FALSE;
			}
		}
	return(new.ids);
	}

### PREPARE SESSION ################################################################################
# import command line arguments
library(argparse);

parser <- ArgumentParser();

parser$add_argument('-p', '--project', type = 'character', help = 'PROJECT name');
parser$add_argument('-o', '--output_dir', type = 'character', help = 'path to output directory');
parser$add_argument('-r', '--report_dir', type = 'character', help = 'path to report directory');
parser$add_argument('-g', '--genome', type = 'character', help = 'reference type', default = 'hg38');
parser$add_argument('-y', '--sample_yaml', type = 'character', help = 'path to sample (BAM) yaml');
parser$add_argument('-c', '--cnas', type = 'character',
	help = 'path to gene by sample matrix of cna calls (generated by collect_panelCN_mops_output.R)');
parser$add_argument('-v', '--snp_dir', type = 'character',
	help = 'path to directory containing germline MAF files');
parser$add_argument('-e', '--somatic', type = 'character', help = 'path to ensemble (somatic) MAF file');
parser$add_argument('-m', '--methylation', type = 'character',
	help = 'path to per-site methylation matix (generated by collect_methyldackel_output.R)');
parser$add_argument('-i', '--ichor', type = 'character', help = 'path to ichorCNA TF estimates');
parser$add_argument('-n', '--ichor_cnas', type = 'character', help = 'path to ichorCNA calls');

arguments <- parser$parse_args();

###
#arguments$project <- 'UHNBB';
#arguments$output_dir <- '/cluster/projects/pughlab/pughlab/projects/MultiMMR/Patient_Reports/SUMMARY';
#arguments$report_dir <- '/cluster/projects/pughlab/pughlab/projects/MultiMMR/Patient_Reports/Reports';
#arguments$ichor <- '/cluster/projects/pughlab/pughlab/projects/MultiMMR/UHNBB/DynaCare/sWGS/IchorCNA/2026-01-20_UHNBB_ichorCNA_estimates.tsv'
#arguments$ichor_cnas <- '/cluster/projects/pughlab/pughlab/projects/MultiMMR/UHNBB/DynaCare/sWGS/IchorCNA/2026-01-20_UHNBB_perbin_cna_status.tsv'
#arguments$sample_yaml <- '/cluster/projects/pughlab/pughlab/projects/MultiMMR/UHNBB/DynaCare/DNASeq/GATK/gatk_bam_config_2025-08-25_15-25-35.yaml'
#arguments$snp_dir <- '/cluster/projects/pughlab/pughlab/projects/MultiMMR/UHNBB/DynaCare/DNASeq/HaplotypeCaller/cohort/VCF2MAF';
#arguments$cnas <- '/cluster/projects/pughlab/pughlab/projects/MultiMMR/UHNBB/DynaCare/DNASeq/panelCNmops/2025-10-29_MultiMMR_panelCN.mops_output.tsv'
#arguments$methylation <- '/cluster/projects/pughlab/pughlab/projects/MultiMMR/UHNBB/DynaCare/EMSeq/MethylDackel/2025-10-10_MultiMMR_CpG_methylation_matrix.RData'
#arguments$somatic <- '/cluster/projects/pughlab/pughlab/projects/MultiMMR/Patient_Reports/data/DNASeq/ENSEMBLE/UHNBB_ensemble_mutation_data.tsv'
###


# load libraries
library(BoutrosLab.plotting.general);
library(GenomicRanges);
library(plyr);
library(yaml);
library(xtable);

# what's the date?
date <- Sys.Date();

# get input files
output.dir <- arguments$output_dir;
reports.dir <- arguments$report_dir;

# create (if necessary) and move to output directory
if (!dir.exists(output.dir)) {
	dir.create(output.dir);
	}

if (!dir.exists(reports.dir)) {
	dir.create(reports.dir);
	}

genes.of.interest <- unlist(strsplit('MLH1|MSH2|MSH3|MSH6|PMS2|APC|POLD1|POLE|MUTYH|TP53|BRAF','\\|'));

### READ DATA ######################################################################################
# get data
if (!is.null(arguments$ichor)) {
	tf.data <- read.delim(arguments$ichor);
	} else {
	tf.data <- NULL;
	}

if (!is.null(arguments$ichor_cnas)) {
	ichor.data <- read.delim(arguments$ichor_cnas);
	} else {
	ichor.data <- NULL;
	}

if (!is.null(arguments$somatic)) {
	mutation.data <- read.delim(arguments$somatic, comment.char = '#');
	} else {
	mutation.data <- NULL;
	}

if (is.null(arguments$cnas)) {
	stop('ERROR: No CNA calls provided, please provide path to CNA matrix from panelCN.mops.');
	} else {
	cna.data <- read.delim(arguments$cnas, stringsAsFactors = FALSE);
	}

if (is.null(arguments$snp_dir)) {
	stop('ERROR: No SNP directory provided, please provide path to variant MAF file(s).');
	} else {
	snp.files <- list.files(path = arguments$snp_dir, pattern = 'maf$', full.names = TRUE);
	}

if (is.null(arguments$methylation)) {
	stop('ERROR: No methylation data provided, please provide path to methylation matrix.');
	} else {
	load(arguments$methylation);
	methylation.data <- full.methylation.data;
	rm(full.methylation.data);
	}

# parse sample information from yaml file
project.yaml <- read_yaml(arguments$sample_yaml);

sample.info <- as.data.frame(matrix(nrow = 0, ncol = 3));
colnames(sample.info) <- c('Patient','Sample','Type');

for (patient in names(project.yaml)) {
	normals <- names(project.yaml[[patient]]$normal);
	tumours <- names(project.yaml[[patient]]$tumour);
	sample.info <- rbind(sample.info, 
		data.frame(Patient = rep(patient, length(normals)), Sample = normals, Type = rep('normal', length(normals))),
		data.frame(Patient = rep(patient, length(tumours)), Sample = tumours, Type = rep('tumour', length(tumours)))
		);
	}

sample.info$Patient <- as.character(sample.info$Patient);
sample.info$Sample <- as.character(sample.info$Sample);

sample.info <- sample.info[order(sample.info$Patient, sample.info$Sample),];
normal.smps <- sample.info[which(sample.info$Type == 'normal'),]$Sample;
tumour.smps <- sample.info[which(sample.info$Type == 'tumour'),]$Sample;


### FORMAT DATA ####################################################################################
# format CNA data
anno.fields <- c('Chr','Start','End','Gene','Exon');

if (any(is.na(cna.data$Gene))) {
	cna.data[is.na(cna.data$Gene),]$Gene <- 'region';
	}

gene.anno <- unique(cna.data[,anno.fields]);
target.regions <- makeGRangesFromDataFrame(gene.anno,
	keep.extra.columns = TRUE,
	seqnames.field = 'Chr',
	start.field = 'Start',
	end.field = 'End');

# calculate log2ratios
print('Formatting panelCN.mops data...');
cna.data$Ratio <- log2(cna.data$RC.norm / cna.data$medRC.norm);

# set minimum ratio to -2 (to deal with -Inf)
cna.data$CN.Abs <- as.numeric(gsub('CN','',cna.data$CN))-2;

# convert to region x sample matrix
cna.matrix <- reshape(
	cna.data[,c('Chr','Start','End','Gene','Exon','Sample','CN.Abs')],
	direction = 'wide',
	idvar = c('Chr','Start','End','Gene','Exon'),
	timevar = 'Sample'
	);
colnames(cna.matrix) <- gsub('CN.Abs.','',colnames(cna.matrix));

cna.matrix$Chr <- factor(cna.matrix$Chr, levels = paste0('chr',1:22));
cna.matrix <- cna.matrix[order(cna.matrix$Chr, cna.matrix$Start),];

# convert to region x sample matrix
ratio.matrix <- reshape(
	cna.data[,c('Chr','Start','End','Gene','Exon','Sample','Ratio')],
	direction = 'wide',
	idvar = c('Chr','Start','End','Gene','Exon'),
	timevar = 'Sample'
	);
colnames(ratio.matrix) <- gsub('Ratio.','',colnames(ratio.matrix));

ratio.matrix$Chr <- factor(ratio.matrix$Chr, levels = paste0('chr',1:22));
ratio.matrix <- ratio.matrix[order(ratio.matrix$Chr, ratio.matrix$Start),];

# get sample ids
cna.samples <- setdiff(colnames(ratio.matrix), anno.fields);

# run segmentation
CNAobj <- DNAcopy::CNA(ratio.matrix[,cna.samples], chrom = ratio.matrix$Chr, maploc = ratio.matrix$Start,
	data.type = 'logratio', sampleid = cna.samples, presorted = TRUE);

seg.data <- DNAcopy::segment(CNAobj)$output;
if (any(!cna.samples %in% seg.data$ID)) {
	seg.data$ID <- gsub('\\.','-',seg.data$ID);
	}


# format germline SNP data
print('Formatting germline SNP data...');
snp.data <- do.call(rbind, lapply(snp.files, read.delim, comment.char = '#'));
snp.data$t_vaf <- snp.data$t_alt_count / snp.data$t_depth;
snp.data$n_vaf <- snp.data$n_alt_count / snp.data$n_depth;
snp.data$Chromosome <- factor(snp.data$Chromosome, levels = paste0('chr',c(1:22,'X')));

# extract tumour VAFs for heterozygous germline variants
baf.data <- do.call(rbind, lapply(tumour.smps, function(smp) {
	t.idx <- which(snp.data$Tumor_Sample_Barcode == smp);
	snp.idx <- which(is.na(snp.data$n_vaf) | (
		snp.data$n_vaf >= 0.35 & snp.data$n_vaf <= 0.65));

	tmp.data <- snp.data[intersect(t.idx, snp.idx),];
	if (nrow(tmp.data) > 0) {
		tmp.data <- tmp.data[order(tmp.data$Chromosome, tmp.data$Start_Position),]; 

		# run segmentation
		SNPobj <- DNAcopy::CNA(tmp.data$t_vaf, chrom = tmp.data$Chromosome, maploc = tmp.data$Start_Position,
			data.type = 'logratio', sampleid = smp, presorted = TRUE);

		output <- DNAcopy::segment(SNPobj)$output;
		return(output);
		}
	}));

baf.data$ID <- gsub('\\.','-', baf.data$ID);


# format somatic SNV/INDEL data
if (!is.null(mutation.data)) {
	print('Formatting somatic SNV/INDEL data...');
	mutation.data <- mutation.data[which(mutation.data$Hugo_Symbol %in% genes.of.interest),];
	mutation.data$ClinVar <- sapply(mutation.data$CLIN_SIG, function(i) {
		if (is.na(i)) { return(NA) 
			} else {
			tmp <- unlist(strsplit(i,','));
			x <- if (any(tmp == 'pathogenic')) { 'pathogenic'
				} else if (any(tmp == 'likely_pathogenic')) { 'likely_pathogenic'
				} else if (any(tmp %in% c('uncertain_significance','conflicting_interpretations_of_pathogenicity'))) { 'VUS'
				} else if (any(tmp == 'likely_benign')) { 'likely_benign'
				} else if (any(tmp == 'benign')) { 'benign'
				} else { NA }
			return(x);
			}
		});

	keep.fields <- c('Tumor_Sample_Barcode','Hugo_Symbol','Chromosome','Start_Position','End_Position',
		'Variant_Classification','Variant_Type','Reference_Allele','Tumor_Seq_Allele2',
		'HGVSc','HGVSp','dbSNP_RS','ClinVar','t_depth','t_alt_count','n_depth','n_alt_count');

	mutation.data <- mutation.data[grepl('pathogenic', mutation.data$ClinVar),keep.fields];
	}


# format methylation data
print('Formatting methylation data...');
m.data <- list();
for (patient in unique(sample.info$Patient)) {
	tumours <- sample.info[which(sample.info$Patient == patient & 
		sample.info$Type == 'tumour'),]$Sample;
	tumours <- intersect(tumours, colnames(methylation.data));
	if (length(tumours) == 0) { next; }

	normals <- sample.info[which(sample.info$Patient == patient &
		sample.info$Type == 'normal'),]$Sample;
	normals <- intersect(normals, colnames(methylation.data));
	if (length(normals) == 0) {
		normals <- intersect(sample.info[which(sample.info$Type == 'normal'),]$Sample,
			colnames(methylation.data));
		}
	if (length(normals) == 0) { next; }

	anno.data <- methylation.data[,1:3];
	n.data <- if (length(normals) > 1) {
		apply(methylation.data[,normals],1,mean,na.rm = TRUE);
		} else { methylation.data[,normals]; }
	t.data <- methylation.data[,tumours] - n.data;

	m.data[[patient]] <- na.omit(cbind(anno.data, t.data));
	colnames(m.data[[patient]]) <- c(colnames(anno.data),tumours);
	}

methyl.data <- join_all(m.data);


# set up colour schemes
colour.scheme <- c('magenta1','black','darkolivegreen1');
colour.function.low <- colorRamp(colour.scheme[1:2], space = 'Lab');
colour.function.high <- colorRamp(colour.scheme[2:3], space = 'Lab');

methyl.colours <- c(
	rgb(colour.function.low(seq(0,0.9,0.1)^(1/0.3)), maxColorValue = 255),
	colour.scheme[2],
	rgb(colour.function.high(seq(0.1,1,0.1)^(0.3)), maxColorValue = 255)
	)
methyl.legend.labels <- rep(NA, length(methyl.colours));
methyl.legend.labels[c(1,11,21)] <- c('-1','0','+1');

colour.scheme <- colorRampPalette(c('blue','grey90','red'));
cna.colours <- colour.scheme(length(seq(-1,1,0.1)));
cna.legend.labels <- rep(NA, length(cna.colours));
cna.legend.labels[c(1,11,21)] <- c('-1','0','+1');

abs.colours <- cna.colours[c(1,6,11,16,21)];
names(abs.colours) <- c('-2','-1','0','1','2');

# move to output directory
setwd(output.dir);

# set up data template
template <- data.frame(slidingWindows(target.regions,1));
template$Exon <- factor(
	template$group,
	levels = c(1:length(target.regions)),
	labels = target.regions$Exon
	);

colnames(template)[3:4] <- c('chromosome','position');
template <- template[,c('chromosome','position','Exon')];

template$Order <- 1:nrow(template);

template <- merge(template, gene.anno[,c('Exon','Gene')]);
template <- template[order(template$Order),];

# prepare data for each sample
sample.data <- list();

for (smp in tumour.smps) {

	print(paste('Processing sample: ', smp));

	# check for data completeness
	if (!smp %in% colnames(cna.matrix)) {
		smp.data <- template;
		smp.data[,c('CN','c.colour')] <- NA;
		} else {

		# format log-ratio values
		smp.data <- merge(
			template, 
			cna.matrix[,c('Chr','Exon',smp)], 
			by.x = c('chromosome','Exon'),
			by.y = c('Chr','Exon'),
			all.x = TRUE
			);
		colnames(smp.data)[which(colnames(smp.data) == smp)] <- 'CN';
		smp.data$c.colour <- abs.colours[match(as.character(smp.data$CN), names(abs.colours))];
		}

	# check for data completeness
	if (!smp %in% colnames(ratio.matrix)) {
		smp.data <- template;
		smp.data[,c('LRR','r.colour')] <- NA;
		} else {

		# format log-ratio values
		smp.data <- merge(
			smp.data, 
			ratio.matrix[,c('Chr','Exon',smp)], 
			by.x = c('chromosome','Exon'),
			by.y = c('Chr','Exon'),
			all.x = TRUE
			);
		colnames(smp.data)[which(colnames(smp.data) == smp)] <- 'LRR';
		smp.data$r.colour <- 'black';
		}

	# check for data completeness
	if (!smp %in% snp.data$Tumor_Sample_Barcode) {
		smp.data[,c('BAF','b.colour')] <- NA;
		} else {

		# format B-allele frequencies
		snp.to.plot <- snp.data[which(snp.data$Tumor_Sample_Barcode == smp),];
		snp.idx <- which(is.na(snp.to.plot$n_vaf) | (
			snp.to.plot$n_vaf >= 0.35 & snp.to.plot$n_vaf <= 0.65));
		smp.data <- merge(
			smp.data,
			snp.to.plot[snp.idx,c('Chromosome','Start_Position','t_vaf')], 
			by.x = c('chromosome','position'),
			by.y = c('Chromosome','Start_Position'),
			all.x = TRUE
			);
		colnames(smp.data)[which(colnames(smp.data) == 't_vaf')] <- 'BAF';
		smp.data$b.colour <- 'black';
		if (any(na.omit(smp.data$BAF) < 0.2 | na.omit(smp.data$BAF) > 0.8)) {
			smp.data[which(smp.data$BAF < 0.2 | smp.data$BAF > 0.8),]$b.colour <- 'orange'; 
			}
		}

	# check for data completeness
	if (!smp %in% baf.data$ID) {
		smp.data[,c('BAF.Seg')] <- NA;
		} else {
		# format CN segment calls
		tmp <- makeGRangesFromDataFrame(smp.data,
			seqnames.field = 'chromosome',
			start.field = 'position',
			end.field = 'position',
			keep.extra.columns = TRUE
			);

		segs <- makeGRangesFromDataFrame(
			baf.data[which(baf.data$ID == smp),],
			seqnames.field = 'chrom',
			start.field = 'loc.start',
			end.field = 'loc.end',
			keep.extra.columns = TRUE
			);

		tmp2 <- suppressWarnings(mergeByOverlaps(tmp, segs));
		df <- data.frame(tmp2$tmp, tmp2$seg.mean);

		smp.data <- merge(
			smp.data,
			df[,c('seqnames','start','tmp2.seg.mean')],
			by.x = c('chromosome','position'),
			by.y = c('seqnames','start'),
			all.x = TRUE
			);
		colnames(smp.data)[which(colnames(smp.data) == 'tmp2.seg.mean')] <- 'BAF.Seg';
		}

	# check for data completeness
	if (!smp %in% mutation.data$Tumor_Sample_Barcode) {
		smp.data[,c('Somatic')] <- NA;
		} else {

		# format somatic mutations
		tmp <- mutation.data[which(mutation.data$Tumor_Sample_Barcode == smp),];
		smp.data <- merge(
			smp.data,
			tmp[,c('Chromosome','Start_Position','HGVSc')],
			by.x = c('chromosome','position'),
			by.y = c('Chromosome','Start_Position'),
			all.x = TRUE
			);
		colnames(smp.data)[which(colnames(smp.data) == 'HGVSc')] <- 'Somatic';
		}

	# check for data completeness
	if (!smp %in% colnames(methyl.data)) {
		smp.data[,c('Methylation','m.colour')] <- NA;
		} else {

		# format methylation levels
		smp.data <- merge(
			smp.data,
			methyl.data[,c('seqnames','start',smp)], 
			by.x = c('chromosome','position'),
			by.y = c('seqnames','start'),
			all.x = TRUE
			);
		colnames(smp.data)[which(colnames(smp.data) == smp)] <- 'Methylation';
		smp.data$m.colour <- methyl.colours[match(as.character(round(smp.data$Methylation,1)), seq(-1,1,0.1))];
		}

	# check for data completeness
	if (!smp %in% seg.data$ID) {
		smp.data[,c('CN.Seg','s.colour')] <- NA;
		} else {
		# format CN segment calls
		tmp <- makeGRangesFromDataFrame(smp.data,
			seqnames.field = 'chromosome',
			start.field = 'position',
			end.field = 'position',
			keep.extra.columns = TRUE
			);

		segs <- makeGRangesFromDataFrame(
			seg.data[which(seg.data$ID == smp),],
			seqnames.field = 'chrom',
			start.field = 'loc.start',
			end.field = 'loc.end',
			keep.extra.columns = TRUE
			);

		tmp2 <- mergeByOverlaps(tmp, segs);
		df <- data.frame(tmp2$tmp, tmp2$seg.mean);

		smp.data <- merge(
			smp.data,
			df[,c('seqnames','start','tmp2.seg.mean')],
			by.x = c('chromosome','position'),
			by.y = c('seqnames','start'),
			all.x = TRUE
			);
		colnames(smp.data)[which(colnames(smp.data) == 'tmp2.seg.mean')] <- 'CN.Seg';
		smp.data$s.colour <- cna.colours[match(as.character(round(smp.data$CN.Seg,1)), seq(-1,1,0.1))];
		if (any(na.omit(smp.data$CN.Seg) < -1)) {
			smp.data[which(smp.data$CN.Seg < -1),]$s.colour <- cna.colours[1];
			}
		if (any(na.omit(smp.data$CN.Seg) > 1)) {
			smp.data[which(smp.data$CN.Seg > 1),]$s.colour <- cna.colours[length(cna.colours)];
			}
		}

	# check for data completeness
	if (!smp %in% ichor.data$ID) {
		smp.data[,c('WGCN.Seg','wg.colour')] <- NA;
		} else {
		# format CN segment calls
		tmp <- makeGRangesFromDataFrame(smp.data,
			seqnames.field = 'chromosome',
			start.field = 'position',
			end.field = 'position',
			keep.extra.columns = TRUE
			);

		segs <- makeGRangesFromDataFrame(
			ichor.data[which(ichor.data$ID == smp),],
			seqnames.field = 'chrom',
			start.field = 'start',
			end.field = 'end',
			keep.extra.columns = TRUE
			);

		tmp2 <- mergeByOverlaps(tmp, segs);
		df <- data.frame(tmp2$tmp, tmp2$seg.median.logR);

		smp.data <- merge(
			smp.data,
			df[,c('seqnames','start','tmp2.seg.median.logR')],
			by.x = c('chromosome','position'),
			by.y = c('seqnames','start'),
			all.x = TRUE
			);
		colnames(smp.data)[which(colnames(smp.data) == 'tmp2.seg.median.logR')] <- 'WGCN.Seg';
		smp.data$wg.colour <- cna.colours[match(as.character(round(smp.data$WGCN.Seg,1)), seq(-1,1,0.1))];
		if (any(na.omit(smp.data$WGCN.Seg) < -1)) {
			smp.data[which(smp.data$WGCN.Seg < -1),]$wg.colour <- cna.colours[1];
			}
		if (any(na.omit(smp.data$WGCN.Seg) > 1)) {
			smp.data[which(smp.data$WGCN.Seg > 1),]$wg.colour <- cna.colours[length(cna.colours)];
			}
		}

	sample.data[[smp]] <- smp.data;
	}

save(
	sample.data,
	file = generate.filename(arguments$project, 'summarized_panel_data', 'RData')
	);

# run unlink, in case it exists from a previous run
link.name <- paste0(arguments$project, '_summarized_panel_data.RData');
unlink(link.name);

file.symlink(
	generate.filename(arguments$project, 'summarized_panel_data', 'RData'),
	link.name
	);


### MAKE PLOTS #####################################################################################
new.axis.function <- function(side, line.col = "black", ...) {
	if (side %in% c("left")) {
		axis.default(side = side, line.col = "black", ...);
		lims <- current.panel.limits();
		panel.abline(h = lims$ylim[1], v = lims$xlim[1]);
		}
	}

# find chromosome breaks
get.midpoints <- function(x) {
	y <- get.line.breaks(x)+0.5;
	(c(1, y) - 1) + ( c(y, length(x)) - c(1, y) ) / 2
	}
chr.breaks <- get.line.breaks(template$chromosome);
chr.midpoints <- get.midpoints(template$chromosome);

# find genes of interest
gene.pos <- aggregate(Order ~ Gene, template[which(template$Gene %in% genes.of.interest),], median);

# create some legends
plot.legends <- list(
	legend = list(
		colours = 'orange',
		labels = 'LOH'
		),
	legend = list(
		title = 'CN Segment',
		colours = c('blue','grey90','red'),
		labels = c('-2','0','2'),
		continuous = TRUE,
		tck = 0.5,
		tck.number = 2,
		width = 5,
		height = 1,
		angle = 1
		),
	legend = list(
		title = 'Methylation',
		colours = c('magenta1','black','darkolivegreen1'),
		labels = c('-1','0','1'),
		continuous = TRUE,
		tck = 0.5,
		tck.number = 2,
		width = 5,
		height = 1,
		angle = 1
		),
	legend = list(
		colours = 'red',
		labels = 'pathogenic variant'
		)
	);

my.legend.grob <- legend.grob(
	legends = plot.legends,
	title.just = 'left',
	label.cex = 0.7,
	title.cex = 0.7,
	size = 1.5,
	between.col = 4,
	layout = c(4,1) 
	);

caption <- "Complete data summary across the MultiMMR panel: BAF = tumour b-allele frequency for heterozygous germline SNPs and/or small INDELs; panelCNA = log2(reads ratio) for exon-level CN calls with segmentation by CBS; ichorCNA = log2(reads ratio) for genome-wide CN calls from sWGS; Methylation = relative methylation levels (vs matched normal or panel of normals where available).";


### PER-SAMPLE SUMMARIES ###########################################################################
setwd(reports.dir);

# initiate report object
tex.file <- 'complete_panel_summary.tex';

for (patient in unique(sample.info$Patient)) {

	print(paste('Generating report files for patient: ', patient));

	if (!dir.exists(patient)) {
		dir.create(patient);
		}

	setwd(patient);

	patient.data <- sample.info[which(sample.info$Patient == patient),];

	# is a matched normal present?
	matched.normal <- if (any(patient.data$Type == 'normal')) {
		patient.data[which(patient.data$Type == 'normal'),]$Sample[1];
		} else { 'panelOfNormals'; }

	patient.data <- patient.data[which(patient.data$Type == 'tumour'),];

	# add data to tex file
	write("\\section{Panel Summary}\n", file = tex.file);

	for (smp in patient.data$Sample) {

		# pull ichorCNA estimates
		tf.key <- NULL;
		if (!is.null(tf.data)) {
			tf <- ifelse (smp %in% tf.data$ID,
				round(tf.data[which(tf.data$ID == smp),]$Tumour.Fraction,2),
				NA);
			
			tf.key <- list(
				text = list(lab = c(
					paste0('Tumour sample: ', smp),
					paste0('Normal used for reference: ', matched.normal),
					paste0('Tumour fraction (ichorCNA) = ', tf)))
				);
			}

		# get gene data for this sample
		smp.data <- sample.data[[smp]];

		# plot SNP frequencies
		baf.plot <- create.scatterplot(
			BAF ~ Order,
			smp.data,
			col = smp.data$b.colour,
			alpha = 0.8,
			cex = 0.8,
			xlimits = c(0,nrow(template)),
			ylimits = c(0,1),
			xat = chr.midpoints,
			xaxis.lab = c(1:20,'21\n','22'),
			yat = seq(0,1,0.5),
			xaxis.tck = c(0.5,0),
			yaxis.tck = c(0.5,0),
			xaxis.cex = 1,
			yaxis.cex = 1,
			xlab.label = NULL,
			ylab.label = 'BAF',
			ylab.cex = 1.5,
			add.rectangle = TRUE,
			xleft.rectangle = c(0,chr.breaks),
			xright.rectangle = c(chr.breaks, nrow(template)),
			ytop.rectangle = 5,
			ybottom.rectangle = -5,
			col.rectangle = c('grey90','white'),
			alpha.rectangle = 0.5,
			add.points = TRUE,
			points.col = 'grey60',
			points.cex = 0.1,
			points.alpha = 0.8,
			points.x = smp.data$Order,
			points.y = smp.data$BAF.Seg,
			style = 'Nature'
			);

		baf.axis.function <- function(side, line.col = "black", ...) {
			if (side %in% c("left","top")) {
				axis.default(side = side, line.col = "black", ...);
				lims <- current.panel.limits();
				panel.abline(h = lims$ylim[2], v = lims$xlim[1]);
				}
			}

		baf.plot$axis <- baf.axis.function;
		baf.plot$x.scales$alternating <- 2;

		# plot LRR values
		lrr.plot <- create.scatterplot(
			LRR ~ Order,
			smp.data,
			col = smp.data$r.colour,
			alpha = 0.8,
			cex = 0.7,
			xlimits = c(0,nrow(template)),
			ylimits = c(-2,2),
			xaxis.lab = NULL,
			yat = seq(-2,2,1),
			xaxis.tck = 0,
			yaxis.tck = c(0.5,0),
			yaxis.cex = 1,
			xlab.label = NULL,
			ylab.label = 'panelCNA',
			ylab.cex = 1.5,
			add.rectangle = TRUE,
			xleft.rectangle = c(0,chr.breaks),
			xright.rectangle = c(chr.breaks, nrow(template)),
			ytop.rectangle = 5,
			ybottom.rectangle = -5,
			col.rectangle = c('grey90','white'),
			alpha.rectangle = 0.5,
			add.grid = TRUE,
			ygrid.at = 0,
			xgrid.at = -1,
			add.points = TRUE,
			points.col = smp.data$s.colour,
			points.cex = 0.3,
			points.alpha = 0.8,
			points.x = smp.data$Order,
			points.y = smp.data$CN.Seg,
			style = 'Nature'
			);

		# plot ichorCNA data
		ichor.plot <- create.scatterplot(
			WGCN.Seg ~ Order,
			smp.data,
			col = smp.data$wg.colour,
			alpha = 0.8,
			cex = 0.7,
			xlimits = c(0,nrow(template)),
			ylimits = c(-2,2),
			xaxis.lab = NULL,
			yat = seq(-2,2,1),
			xaxis.tck = 0,
			yaxis.tck = c(0.5,0),
			yaxis.cex = 1,
			xlab.label = NULL,
			ylab.label = 'ichorCNA',
			ylab.cex = 1.5,
			add.rectangle = TRUE,
			xleft.rectangle = c(0,chr.breaks),
			xright.rectangle = c(chr.breaks, nrow(template)),
			ytop.rectangle = 5,
			ybottom.rectangle = -5,
			col.rectangle = c('grey90','white'),
			alpha.rectangle = 0.5,
			add.grid = TRUE,
			ygrid.at = 0,
			xgrid.at = -1,
			style = 'Nature'
			);

		# plot methylation levels
		methyl.plot <- create.scatterplot(
			Methylation ~ Order,
			smp.data,
			col = smp.data$m.colour,
			alpha = 0.8,
			cex = 0.8,
			xlimits = c(0,nrow(template)),
			ylimits = c(-1,1),
			yat = seq(-1,1,1),
			xaxis.tck = 0,
			yaxis.tck = c(0.5,0),
			yaxis.cex = 1,
			xaxis.fontface = 'italic',
			xlab.label = NULL,
			ylab.label = 'Methylation',
			ylab.cex = 1.5,
			add.rectangle = TRUE,
			xleft.rectangle = c(0,chr.breaks),
			xright.rectangle = c(chr.breaks, nrow(template)),
			ytop.rectangle = 5,
			ybottom.rectangle = -5,
			col.rectangle = c('grey90','white'),
			alpha.rectangle = 0.5,
			add.grid = TRUE,
			ygrid.at = 0,
			xgrid.at = -1,
			style = 'Nature'
			);

		lrr.plot$axis <- new.axis.function;
		ichor.plot$axis <- new.axis.function;
		methyl.plot$axis <- new.axis.function;

		# make gene plot with legend
		smp.data$roi <- as.numeric(smp.data$Gene %in% genes.of.interest);

		gene.covariates <- create.scatterplot(
			roi ~ Order,
			smp.data,
			type = 'h',
			cex = 0,
			xlimits = c(0,nrow(smp.data)),
			ylimits = c(0,1),
			xlab.label = NULL,
			ylab.label = NULL,
			xat = gene.pos$Order,
			xaxis.lab = gene.pos$Gene,
			xaxis.rot = 45,
			xaxis.cex = 0.8,
			yaxis.lab = NULL,
			xaxis.tck = 0,
			yaxis.tck = 0,
			abline.v = smp.data[!is.na(smp.data$Somatic),]$Order,
			abline.col = 'red',
			abline.lwd = 2,
			legend = list(inside = list(fun = my.legend.grob, x = 0, y = -3.5)),
			style = 'Nature'
			);

		# combine them
		create.multipanelplot(
			plot.objects = list(baf.plot, lrr.plot, ichor.plot, methyl.plot, gene.covariates),
			layout.width = 1,
			layout.height = 5,
			plot.objects.heights = c(1.1,1,1,1,0.6),
			plot.objects.widths = 1,
			left.legend.padding = 0,
			right.legend.padding = 0,
			top.legend.padding = 0,
			bottom.legend.padding = 0,
			bottom.padding = 3,
			y.spacing = c(-2.5,-1.5,-1.5,-1.5,-1.5),
			x.spacing = 0,
			ylab.axis.padding = 1,
			legend = list(
				inside = list(fun = draw.key, args = list(key = tf.key), x = 0.85, y = 0.06)
				),
			height = 8,
			width = 16,
			resolution = 200,
			filename = generate.filename(smp, '_panel_summary', 'png')
			);

		# add data to tex file
		write("\\begin{figure}[h!]", file = tex.file, append = TRUE);
		write("\\begin{center}", file = tex.file, append = TRUE);
		write(paste0(
			"\\includegraphics[width=1\\textwidth]{",
			generate.filename(smp, '_panel_summary','png'), '}'
			), file = tex.file, append = TRUE);
		write("\\end{center}", file = tex.file, append = TRUE);
		write(paste0(
			"\\caption{", caption, "}"
			), file = tex.file, append = TRUE);
		write("\\end{figure}\n", file = tex.file, append = TRUE);
		write("\\pagebreak\n", file = tex.file, append = TRUE);
		}

	setwd(reports.dir);
	}

### SAVE SESSION INFO ##############################################################################
setwd(output.dir);
save.session.profile(generate.filename(arguments$project, 'PlotPanelSummary_SessionProfile','txt'));
